File: broker.go/internal/stores/sqlite/metrics_store.go
package sqlite

import (
	"context"
	"database/sql"
	"errors"
	"time"

	"monstermq.io/edge/internal/stores"
)

const metricsTable = "metrics"

type MetricsStore struct {
	db *DB
}

func NewMetricsStore(db *DB) *MetricsStore { return &MetricsStore{db: db} }

func (m *MetricsStore) Close() error { return nil }

func (m *MetricsStore) EnsureTable(ctx context.Context) error {
	stmts := []string{
		`CREATE TABLE IF NOT EXISTS ` + metricsTable + ` (
            timestamp TEXT NOT NULL,
            metric_type TEXT NOT NULL,
            identifier TEXT NOT NULL,
            metrics TEXT NOT NULL,
            PRIMARY KEY (timestamp, metric_type, identifier)
        )`,
		`CREATE INDEX IF NOT EXISTS metrics_timestamp_idx ON ` + metricsTable + ` (timestamp)`,
		`CREATE INDEX IF NOT EXISTS metrics_type_identifier_idx ON ` + metricsTable + ` (metric_type, identifier, timestamp)`,
	}
	for _, q := range stmts {
		if _, err := m.db.Exec(q); err != nil {
			return err
		}
	}
	return nil
}

func (m *MetricsStore) StoreMetrics(ctx context.Context, kind stores.MetricKind, ts time.Time, identifier, payload string) error {
	_, err := m.db.Exec(`INSERT OR REPLACE INTO `+metricsTable+` (timestamp, metric_type, identifier, metrics) VALUES (?, ?, ?, ?)`,
		ts.UTC().Format(time.RFC3339Nano), string(kind), identifier, payload)
	return err
}

func (m *MetricsStore) GetLatest(ctx context.Context, kind stores.MetricKind, identifier string) (time.Time, string, error) {
	row := m.db.Conn().QueryRowContext(ctx,
		`SELECT timestamp, metrics FROM `+metricsTable+` WHERE metric_type = ? AND identifier = ? ORDER BY timestamp DESC LIMIT 1`,
		string(kind), identifier)
	var (
		tsStr   string
		payload string
	)
	if err := row.Scan(&tsStr, &payload); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return time.Time{}, "", nil
		}
		return time.Time{}, "", err
	}
	t, _ := time.Parse(time.RFC3339Nano, tsStr)
	return t, payload, nil
}

func (m *MetricsStore) GetHistory(ctx context.Context, kind stores.MetricKind, identifier string, from, to time.Time, limit int) ([]stores.MetricsRow, error) {
	if limit <= 0 {
		limit = 1000
	}
	rows, err := m.db.Conn().QueryContext(ctx,
		`SELECT timestamp, metrics FROM `+metricsTable+` WHERE metric_type = ? AND identifier = ? AND timestamp >= ? AND timestamp <= ? ORDER BY timestamp ASC LIMIT ?`,
		string(kind), identifier, from.UTC().Format(time.RFC3339Nano), to.UTC().Format(time.RFC3339Nano), limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []stores.MetricsRow{}
	for rows.Next() {
		var (
			tsStr   string
			payload string
		)
		if err := rows.Scan(&tsStr, &payload); err != nil {
			return nil, err
		}
		t, _ := time.Parse(time.RFC3339Nano, tsStr)
		out = append(out, stores.MetricsRow{Timestamp: t, Payload: payload})
	}
	return out, rows.Err()
}

func (m *MetricsStore) PurgeOlderThan(ctx context.Context, t time.Time) (int64, error) {
	res, err := m.db.Exec(`DELETE FROM `+metricsTable+` WHERE timestamp < ?`, t.UTC().Format(time.RFC3339Nano))
	if err != nil {
		return 0, err
	}
	n, _ := res.RowsAffected()
	return n, nil
}

File: broker.rs/src/stores/sqlite/metrics_store.rs
use super::db::Db;
use crate::stores::traits::{MetricsStore, StoreResult};
use crate::stores::types::{MetricKind, MetricsRow};
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use sqlx::Row;

pub struct SqliteMetricsStore { db: Db }

impl SqliteMetricsStore { pub fn new(db: Db) -> Self { Self { db } } }

#[async_trait]
impl MetricsStore for SqliteMetricsStore {
    async fn ensure_table(&self) -> StoreResult<()> {
        sqlx::query(r#"CREATE TABLE IF NOT EXISTS metrics (
            timestamp TEXT NOT NULL,
            metric_type TEXT NOT NULL,
            identifier TEXT NOT NULL,
            metrics TEXT NOT NULL,
            PRIMARY KEY (timestamp, metric_type, identifier)
        )"#).execute(&self.db.pool).await?;
        sqlx::query("CREATE INDEX IF NOT EXISTS metrics_timestamp_idx ON metrics (timestamp)").execute(&self.db.pool).await?;
        sqlx::query("CREATE INDEX IF NOT EXISTS metrics_type_identifier_idx ON metrics (metric_type, identifier, timestamp)").execute(&self.db.pool).await?;
        Ok(())
    }

    async fn store_metrics(&self, kind: MetricKind, ts: DateTime<Utc>, identifier: &str, payload: &str) -> StoreResult<()> {
        sqlx::query("INSERT OR REPLACE INTO metrics (timestamp, metric_type, identifier, metrics) VALUES (?, ?, ?, ?)")
            .bind(ts.to_rfc3339_opts(chrono::SecondsFormat::Nanos, true))
            .bind(kind.as_str()).bind(identifier).bind(payload)
            .execute(&self.db.pool).await?;
        Ok(())
    }

    async fn get_latest(&self, kind: MetricKind, identifier: &str) -> StoreResult<Option<(DateTime<Utc>, String)>> {
        let row = sqlx::query("SELECT timestamp, metrics FROM metrics WHERE metric_type = ? AND identifier = ? ORDER BY timestamp DESC LIMIT 1")
            .bind(kind.as_str()).bind(identifier).fetch_optional(&self.db.pool).await?;
        Ok(row.map(|r| {
            let ts_s: String = r.try_get("timestamp").unwrap_or_default();
            let payload: String = r.try_get("metrics").unwrap_or_default();
            let ts = DateTime::parse_from_rfc3339(&ts_s).map(|t| t.with_timezone(&Utc)).unwrap_or_else(|_| Utc::now());
            (ts, payload)
        }))
    }

    async fn get_history(&self, kind: MetricKind, identifier: &str, from: DateTime<Utc>, to: DateTime<Utc>, limit: i64) -> StoreResult<Vec<MetricsRow>> {
        let limit = if limit <= 0 { 1000 } else { limit };
        let rows = sqlx::query("SELECT timestamp, metrics FROM metrics WHERE metric_type = ? AND identifier = ? AND timestamp >= ? AND timestamp <= ? ORDER BY timestamp ASC LIMIT ?")
            .bind(kind.as_str()).bind(identifier)
            .bind(from.to_rfc3339_opts(chrono::SecondsFormat::Nanos, true))
            .bind(to.to_rfc3339_opts(chrono::SecondsFormat::Nanos, true))
            .bind(limit).fetch_all(&self.db.pool).await?;
        Ok(rows.into_iter().map(|r| {
            let ts_s: String = r.try_get("timestamp").unwrap_or_default();
            let payload: String = r.try_get("metrics").unwrap_or_default();
            let ts = DateTime::parse_from_rfc3339(&ts_s).map(|t| t.with_timezone(&Utc)).unwrap_or_else(|_| Utc::now());
            MetricsRow { timestamp: ts, payload }
        }).collect())
    }

    async fn purge_older_than(&self, t: DateTime<Utc>) -> StoreResult<i64> {
        let r = sqlx::query("DELETE FROM metrics WHERE timestamp < ?")
            .bind(t.to_rfc3339_opts(chrono::SecondsFormat::Nanos, true))
            .execute(&self.db.pool).await?;
        Ok(r.rows_affected() as i64)
    }
}
